src/libhook/hooks/syscall.hpp

```diff
@@ -0,0 +1,210 @@
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************/
+#pragma once
+
+#include <string>
+#include "base.hpp"
+#include "../call_result.hpp"
+
+namespace libhook
+{
+
+class SyscallHook : public BaseHook
+{
+public:
+    /**
+     * Factory function to create the trap and perform hooking at the same time.
+     */
+    template<typename Params = CallResult>
+    [[nodiscard]]
+    static auto create(drakvuf_t, const std::string& syscall_name, cb_wrapper_t cb)
+    -> std::unique_ptr<SyscallHook>;
+
+    /**
+     * unhook on dctor
+     */
+    ~SyscallHook() override;
+
+    /**
+     * delete copy ctor, as this class has ownership via RAII
+     */
+    SyscallHook(const SyscallHook&) = delete;
+
+    /**
+     * move ctor, required for move semantics to work properly
+     * important to be noexcept, otherwise bad things will happen
+     */
+    SyscallHook(SyscallHook&&) noexcept;
+
+    /**
+     * delete copy assignment operator, as this class has ownership via RAII
+     */
+    SyscallHook& operator=(const SyscallHook&) = delete;
+
+    /**
+     * move assignment operator, required for move semantics to work properly
+     * important to be noexcept, otherwise bad things will happen
+     */
+    SyscallHook& operator=(SyscallHook&&) noexcept;
+
+    std::string syscall_name_;
+    cb_wrapper_t callback_;
+    drakvuf_trap_t* trap_;
+
+protected:
+    /**
+     * Hide ctor from users, as we enforce factory function usage.
+     */
+    SyscallHook(drakvuf_t, const std::string& syscall_name, cb_wrapper_t cb);
+};
+
+template<typename Params>
+auto SyscallHook::create(drakvuf_t drakvuf, const std::string& syscall_name, cb_wrapper_t cb)
+-> std::unique_ptr<SyscallHook>
+{
+    PRINT_DEBUG("[LIBHOOK] creating syscall hook\n");
+
+    auto hook = std::unique_ptr<SyscallHook>(new SyscallHook(drakvuf, syscall_name, cb));
+    hook->trap_ = new drakvuf_trap_t;
+
+    if (!drakvuf_get_kernel_symbol_rva(hook->drakvuf_, hook->syscall_name_.c_str(), &hook->trap_->breakpoint.rva))
+    {
+        PRINT_DEBUG("[LIBHOOK] Failed to receive addr of function %s\n", hook->syscall_name_.c_str());
+        delete hook->trap_;
+        return std::unique_ptr<SyscallHook>();
+    }
+
+    hook->trap_->breakpoint.lookup_type = LOOKUP_PID;
+    hook->trap_->breakpoint.pid = 4;
+    hook->trap_->breakpoint.addr_type = ADDR_RVA;
+    hook->trap_->breakpoint.module = "ntoskrnl.exe";
```

Comment:
This is all Windows specific hard-coded. At the minimum add a check to ensure the target VM is actually Windows.